src/hotspot/share/asm/register.hpp

```diff
@@ -93,67 +93,75 @@ template <class RegImpl> class ReverseRegSetIterator;
 // A set of registers
 template <class RegImpl>
 class AbstractRegSet {
-  uint32_t _bitset;
+  size_t _bitset;
 
-  AbstractRegSet(uint32_t bitset) : _bitset(bitset) { }
+  constexpr AbstractRegSet(size_t bitset) : _bitset(bitset) { }
+
+  static constexpr int max_size() {
+    return (int)(sizeof(_bitset) * BitsPerByte);
+  }
 
 public:
 
-  AbstractRegSet() : _bitset(0) { }
+  constexpr AbstractRegSet() : _bitset(0) { }
 
-  AbstractRegSet(RegImpl r1) : _bitset(1 << r1->encoding()) { }
+  constexpr AbstractRegSet(RegImpl r1)
+    : _bitset(r1->is_valid() ? size_t(1) << r1->encoding() : 0) { }
```

Comment:
If my assumption from above is correct, we never noticed ppc being broken since _bitset would have been 0 if encoding > 31.

Could you add something like:

```
STATIC_ASSERT(RegImpl::number_of_registers <= 64);
```

?

Reply:
OK.